commonmark/src/Commonmark/Inlines.hs

```diff
@@ -74,18 +74,22 @@ mkInlineParser :: (Monad m, IsInline a)
 mkInlineParser bracketedSpecs formattingSpecs ilParsers attrParsers rm toks = do
   let iswhite t = hasType Spaces t || hasType LineEnd t
   let attrParser = choice attrParsers
-  let toks' = dropWhile iswhite . reverse . dropWhile iswhite . reverse $ toks
-  res <- {-# SCC parseChunks #-} evalStateT
-          (parseChunks bracketedSpecs formattingSpecs ilParsers
-           attrParser rm toks') defaultEnders
+  let go chunks toks' bottoms = do
+        chunks' <- {-# SCC parseChunks #-} evalStateT
+                (parseChunks bracketedSpecs formattingSpecs ilParsers
+                 attrParser rm toks') defaultEnders
+        case chunks' of
+           Left err -> return $ Left err
+           Right chunks'' ->
+              case (processBrackets bracketedSpecs rm (chunks ++ chunks'') bottoms) of
+                  Left st -> go ((reverse . befores . rightCursor) st) (mconcat (map chunkToks $ (maybeToList . center $ rightCursor st) ++ (afters $ rightCursor st))) (stackBottoms st)
+                  Right chunks''' -> return $ Right chunks'''
+  res <- go [] ((dropWhile iswhite . reverse . dropWhile iswhite . reverse) toks) mempty
   return $!
-    case res of
-       Left err     -> Left err
-       Right chunks ->
-         (Right .
-          unChunks .
-          processEmphasis .
-          processBrackets bracketedSpecs rm) chunks
+      case res of
+         Left err     -> Left err
+         Right chunks ->
+           (Right . unChunks . processEmphasis) chunks
```

Comment:
This has gotten pretty complex, and I have a hard time following what is going on.  I'm afraid I'll be at a loss if I ever need to modify this part of the code again. Can it be made a bit more user-friendly, with some comments and simplifying definitions?